File: smallpict-python/app/routes/route.py
from typing import Optional
from app.config import settings
from fastapi import APIRouter, status, Depends, UploadFile, File, Form, Request, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.helper.smallpict import upload
from datetime import datetime
from fastapi.responses import JSONResponse
from app.database import check_db_connection

router = APIRouter()

@router.get("/")
def get_base_status():
    return {
        "status": "ok",
        "service": settings.APP_NAME,
        "version": "1.0.0"
    }

@router.get("/health")
def health_check():
    db_ok = check_db_connection()
    status_code = status.HTTP_200_OK if db_ok else status.HTTP_503_SERVICE_UNAVAILABLE

    return JSONResponse(
        status_code=status_code,
        content={
            "status": "ok" if db_ok else "unhealthy",
            "database": "connected" if db_ok else "disconnected",
            "timestamp": datetime.utcnow().isoformat()
        }
    )

@router.post("/upload")
async def handle_upload(
    request: Request,
    db: Session = Depends(get_db),
    file: Optional[UploadFile] = File(None),
    image: Optional[UploadFile] = File(None),
    name: Optional[str] = Form(None),
    filename: Optional[str] = Form(None)
):
    content_type = request.headers.get("content-type", "").lower()

    if "application/json" in content_type:
        try:
            body = await request.json()
        except Exception:
            raise HTTPException(status_code=400, detail="Invalid JSON payload")

        base64_str = body.get("image") or body.get("base64") or body.get("data")
        file_name = body.get("name") or body.get("filename")

        if not base64_str:
            raise HTTPException(status_code=400, detail="Field 'image' or 'base64' is required in JSON payload")

        try:
            img_record = await upload(
                db=db,
                base64_str=base64_str,
                original_filename=file_name
            )
            return {
                "status": "success",
                "message": "Image uploaded successfully",
                "data": img_record.to_dict()
            }
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    upload_target = file or image
    if upload_target:
        file_bytes = await upload_target.read()
        target_filename = name or filename or upload_target.filename
        try:
            img_record = await upload(
                db=db,
                file_bytes=file_bytes,
                original_filename=target_filename
            )
            return {
                "status": "success",
                "message": "Image uploaded successfully",
                "data": img_record.to_dict()
            }
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    form = await request.form()
    base64_str = form.get("image") or form.get("base64")
    if base64_str and isinstance(base64_str, str):
        target_filename = form.get("name") or form.get("filename")
        try:
            img_record = await upload(
                db=db,
                base64_str=base64_str,
                original_filename=target_filename
            )
            return {
                "status": "success",
                "message": "Image uploaded successfully",
                "data": img_record.to_dict()
            }
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    raise HTTPException(status_code=400, detail="No file or image payload found in request")

File: smallpict-js/src/routes/route.ts
import { Router, Request, Response } from 'express';
import multer from 'multer';
import { config } from '../config/config.js';
import { checkDB } from '../database/db.js';
import { processUpload } from '../helper/smallpict.js';

const router = Router();
const upload = multer({ storage: multer.memoryStorage() });

const health = async (req: Request, res: Response): Promise<void> => {
  try {
    const dbOk = await checkDB();
    const statusCode = dbOk ? 200 : 503;

    res.status(statusCode).json({
      status: dbOk ? 'ok' : 'unhealthy',
      database: dbOk ? 'connected' : 'disconnected',
      timestamp: new Date().toISOString(),
    });
  } catch (err: any) {
    res.status(503).json({
      status: 'unhealthy',
      database: 'disconnected',
      error: err.message,
      timestamp: new Date().toISOString(),
    });
  }
}

const uploadHandler = async (req: Request, res: Response): Promise<void> => {
  try {
    const contentType = req.headers['content-type'] || '';

    if (contentType.includes('application/json')) {
      const base64Str = req.body?.image || req.body?.base64 || req.body?.data;
      const customFilename = req.body?.name || req.body?.filename;

      if (!base64Str) {
        res.status(400).json({
          status: 'error',
          message: "Field 'image' or 'base64' is required in JSON payload",
        });
        return;
      }

      const record = await processUpload(undefined, base64Str, customFilename);
      res.json({
        status: 'success',
        message: 'Image uploaded successfully',
        data: record,
      });
      return;
    }

    const files = req.files as Express.Multer.File[] | undefined;
    const file = (files && files.length > 0) ? files[0] : req.file;
    const customFilename = req.body?.name || req.body?.filename;
    const base64Form = req.body?.base64 || req.body?.image;

    if (!file && !base64Form) {
      res.status(400).json({
        status: 'error',
        message: 'No file or base64 image provided in request',
      });
      return;
    }

    const record = await processUpload(
      file?.buffer,
      base64Form,
      customFilename || file?.originalname,
      file?.mimetype
    );

    res.json({
      status: 'success',
      message: 'Image uploaded successfully',
      data: record,
    });
  } catch (err: any) {
    res.status(400).json({
      status: 'error',
      message: err.message || 'Upload failed',
    });
  }
};

router.get('/', (req, res) => {
  res.json({
    status: 'ok',
    service: config.appName,
    version: '1.0.0',
  });
});
router.get('/health', health);
router.post('/upload', upload.any(), uploadHandler);

export default router;
